src/util/merkleTree.ts

```diff
@@ -0,0 +1,150 @@
+/**
+ * @module util/merkleTree
+ */
+
+import assert from "assert";
+import {bytes32} from "../types";
+import {hash} from "../util/crypto";
+import {intDiv} from "../util/math";
+
+interface IProgressiveMerkleTree {
+  /**
+   * The number of items in the tree
+   */
+  count(): number;
+  depth(): number;
+  /**
+   * Add a new item into the tree
+   *
+   * Return a proof to verify the item is in the tree
+   * @returns proof
+   */
+  push(item: bytes32): bytes32[];
+  /**
+   * The merkle root of the tree
+   */
+  root(): bytes32;
+}
+
+export class ProgressiveMerkleTree implements IProgressiveMerkleTree {
+  private _depth: number;
+  private _count: number;
+  private _branch: bytes32[];
+  private _zerohashes: bytes32[];
+
+  public constructor(depth: number) {
+    assert(depth <= 52, "tree depth must be less than 53");
+    this._depth = depth;
+    this._count = 0;
+    this._branch = Array.from({length: depth}, () => Buffer.alloc(32));
+    this._zerohashes = Array.from({length: depth}, () => Buffer.alloc(32));
+    for (let i = 0; i < depth - 1; i++) {
+      this._zerohashes[i + 1] = this._branch[i + 1] =
```

Comment:
In the pyspec they do `[i - 1]`

Reply:
they go from 1 to depth, we go from 0 to depth - 1